src/query/phrase_prefix_query/phrase_prefix_query.rs

```diff
@@ -0,0 +1,160 @@
+    }
+
+    /// The [`Field`] this `PhrasePrefixQuery` is targeting.
+    pub fn field(&self) -> Field {
+        self.field
+    }
+
+    /// `Term`s in the phrase without the associated offsets.
+    pub fn phrase_terms(&self) -> Vec<Term> {
+        // TODO should we include the last term too?
+        self.phrase_terms
+            .iter()
+            .map(|(_, term)| term.clone())
+            .collect::<Vec<Term>>()
+    }
+
+    /// Returns the [`PhrasePrefixWeight`] for the given phrase query given a specific `searcher`.
+    ///
+    /// This function is the same as [`Query::weight()`] except it returns
+    /// a specialized type [`PhraseQueryWeight`] instead of a Boxed trait.
+    /// If the query was only one term long, this returns `None` wherease [`Query::weight`]
+    /// returns a boxed [`RangeWeight`]
+    pub(crate) fn phrase_query_weight(
+        &self,
+        enable_scoring: EnableScoring<'_>,
+    ) -> crate::Result<Option<PhrasePrefixWeight>> {
+        if self.phrase_terms.is_empty() {
+            return Ok(None);
+        }
+        let schema = enable_scoring.schema();
+        let field_entry = schema.get_field_entry(self.field);
+        let has_positions = field_entry
+            .field_type()
+            .get_index_record_option()
+            .map(IndexRecordOption::has_positions)
+            .unwrap_or(false);
+        if !has_positions {
+            let field_name = field_entry.name();
+            return Err(crate::TantivyError::SchemaError(format!(
+                "Applied phrase query on field {:?}, which does not have positions indexed",
+                field_name
+            )));
+        }
+        let terms = self.phrase_terms();
+        let bm25_weight_opt = match enable_scoring {
+            EnableScoring::Enabled(searcher) => Some(Bm25Weight::for_terms(searcher, &terms)?),
+            EnableScoring::Disabled { .. } => None,
+        };
+        let weight = PhrasePrefixWeight::new(
+            self.phrase_terms.clone(),
+            self.prefix.clone(),
+            bm25_weight_opt,
+            self.max_expansions,
+        );
+        Ok(Some(weight))
+    }
+}
+
+impl Query for PhrasePrefixQuery {
+    /// Create the weight associated with a query.
+    ///
+    /// See [`Weight`].
+    fn weight(&self, enable_scoring: EnableScoring<'_>) -> crate::Result<Box<dyn Weight>> {
+        if let Some(phrase_weight) = self.phrase_query_weight(enable_scoring)? {
+            Ok(Box::new(phrase_weight))
+        } else {
+            let end_term = if let Some(end_value) = prefix_end(&self.prefix.1.value_bytes()) {
+                let mut end_term = Term::with_capacity(end_value.len());
+                end_term.set_field_and_type(self.field, self.prefix.1.typ());
+                end_term.append_bytes(&end_value);
+                Bound::Excluded(end_term)
+            } else {
+                Bound::Unbounded
+            };
+
+            RangeQuery::new_term_bounds(
```

Comment:
I think we lost the limit in the number of term to expand to which could be catastrophic.
(think about a use typing "t" and hitting enter)

Can you open an issue and auto-attribute it to you?

Reply:
opened #1886